src/dune_action_plugin/examples/simple_concatenation.ml

```diff
@@ -0,0 +1,11 @@
+open Dune_action_plugin
+
+let action =
+  let source1 = read_file ~path:(Path.of_string "source1")
+  and source2 = read_file ~path:(Path.of_string "source2") in
+  both source1 source2
+  |> stage ~f:(fun (source1, source2) ->
```

Comment:
This example is rather unfortunate since it doesn't seem to fundamentally require staging: all dependencies are known statically and all you do is manipulate file contents in a manner oblivious to the actual values. This makes me wonder if you should also provide the following function in the API:

```ocaml
val write_file_dyn : path:Path.t -> data:string t -> unit t
```

(Not sure if the `_dyn` suffix is appropriate here, there may be better names.)

Having this function allows one to use the applicative interface here, avoiding the costly `stage`.

An interesting variation of the above `write_file_dyn` is

```
val write_file_dyn_reuse : path:Path.t -> data:string t -> data:string t
```

This allows you to write a value into a file and also reuse it later (e.g. to write it into another file).

Reply:
We were thinking of something like this:

`val write_file_deferred : path:Path.t -> (data:string -> unit) t`

which would be even more flexible.

However, neither of this changes the number of nontrivial stages (where I define a stage to be nontrivial if it has any dependencies), and it seems that it's only the number of nontrivial stages that determines the number of times dune has to run the action.

So we thought this wasn't a very important distinction to make.

Reply:
Hmm, are you saying that `choosing_dependency.ml` has two non-trivial stages, whereas this example has only one, even thought they both use one `stage`? Yes, I can believe this.

If the number of non-trivial stages is the only parameter to care about than the current API seems fine. I think it would be useful to elaborate on the cost model of staging a bit more, to include the notion of "non-trivial stages". As a user of this API I would be very interested to know that the `stage` in this example is essentially free.

One aspect in which `write_file_dyn` and friends seem better is parallelism: if you use a trivial `stage`, you still lose parallelism if you do a sequence `read_file; write_file; read_file`, but you could perform the reads/builds in parallel for `read_file; write_file_dyn; read_file` for some definition of `;`.

Reply:
Yeah, that's right. Added a comment.

I don't understand the example though.

Reply:
Looks like my example was wrong anyway :)